lib/OpenQA/WebAPI/Controller/Test.pm

```diff
@@ -569,14 +569,27 @@ sub prepare_job_results {
       = $schema->resultset('TestSuites')->search({name => \%desc_args}, {columns => [qw(name description)]});
     my %descriptions = map { $_->name => $_->description } @descriptions;
 
-    foreach my $job (@$jobs) {
-        next if $states         && !$states->{$job->state};
-        next if $results        && !$results->{$job->result};
-        next if $archs          && !$archs->{$job->ARCH};
-        next if $machines       && !$machines->{$job->MACHINE};
-        next if $failed_modules && $job->result ne OpenQA::Jobs::Constants::FAILED;
-
-        my $result = $job->overview_result($comment_data, $aggregated, $failed_modules, $self->param('todo')) or next;
+    my @wanted_jobs = grep {
+            (not $states         or $states->{$_->state})
+        and (not $results        or $results->{$_->result})
+        and (not $archs          or $archs->{$_->ARCH})
+        and (not $machines       or $machines->{$_->MACHINE})
+        and (not $failed_modules or $_->result eq OpenQA::Jobs::Constants::FAILED)
+    } @$jobs;
+    my @jobids = map { $_->id } @wanted_jobs;
```

Comment:
Would it be possible to combine the `grep` and `map`? When I remember correctly, `map` allows one to omit elements by returning an empty list.